### PR TITLE
App catalog page, CardTable component and some refactoring

### PR DESCRIPTION
**Pull request by xantorres**
Checklist:
- [x] Fix api calls for Charts
- [x] Basic App Catalog page
- [x] Improve CardTable component with pagination, filtering and sorting

image

### CODE REVIEW

File: src/app/core/components/appCatalog/ApplicationCard.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    padding: theme.spacing.unit
+  }
+})
+
+class ApplicationCard extends React.Component {
+  handleAddToEnv = () => {
+    this.props.handleAddToEnv()
+  }
+
+  handleDeploy = () => {
```
[the remark this answers is not in the document]
Comment (brennancheung) [reply src/app/core/components/appCatalog/ApplicationCard.js]: Also make sure these are required in the `propTypes`.

File: src/app/plugins/kubernetes/components/apps/actions.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import { flatten, prop } from 'ramda'
+
+export const loadClusters = async ({ data, context, setContext, reload }) => {
```
[the remark this answers is not in the document]
Comment (xantorres) [reply src/app/plugins/kubernetes/components/apps/actions.js]: actually, I'm not using this function anymore, so I will remove it

File: src/app/core/components/appCatalog/ApplicationCard.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (xantorres) [reply src/app/core/components/appCatalog/ApplicationCard.js]: yes, adding some logic soon

File: src/server/presets/dev.js [hunk 3]
```diff
@@ -124,37 +126,41 @@ function loadPreset () {
...
-
-  // Nodes
-  // Nodes must be linked to a resMgrHost id or else the UI will break
-  const node = Node.create({
-    data: { name: 'fakeNode1', api_responding: 1, isMaster: 1, uuid: resMgrHost.id },
-    context,
-    raw: true
-  })
-  Node.create({ data: { name: 'fakeNode2', uuid: resMgrHost2.id }, context })
+  range(1, 4).forEach(clusterIdx => {
```
[the remark this answers is not in the document]
Comment (brennancheung) [reply src/server/presets/dev.js]: Slightly unrelated.. I'm thinking we may want to split up the presets.  I'm thinking of having the Cypress tests drive the simulator to create different configs for different scenarios so having functions that we can delegate to to create different configs would work well.  Might also work well while we are developing too.

File: src/app/core/components/cardTable/CardTable.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+      orderBy: null,
+      direction: 'asc',
+      page: 0,
+      selected: [],
+      searchTerm: '',
+      filterValues: {}
+    }
+  }
+
+  handleSearch = value => {
```
Comment (brennancheung) [inline src/app/core/components/cardTable/CardTable.js]: Optional if you like it better.  I know the JS community in general hates "abusing" the short-circuit operations but I find it much easier as a 1-liner than 6 lines.

`handleSearch = searchTerm => this.props.searchTarget && this.setState({ searchTerm })`

File: src/app/core/components/cardTable/CardTable.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+      'sortWith',
+      sorting.find(propEq('id', orderBy))
+    )
+    const sortedRows = [...data].sort((a, b) =>
+      sortWith(b[orderBy], a[orderBy])
+    )
+    return this.state.direction === 'desc' ? sortedRows : sortedRows.reverse()
+  }
+
+  paginate = data => {
```
Comment (brennancheung) [inline src/app/core/components/cardTable/CardTable.js]: We probably want an HOC for pagination but we can do it in another PR.

File: src/app/plugins/kubernetes/components/apps/AppCatalogPage.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    },
+    {
+      field: 'relationships.latestChartVersion.data.created',
+      label: 'Created',
+      sortWith: (prevDate, nextDate) =>
+        moment(prevDate).isBefore(nextDate) ? 1 : -1
+    }
+  ]
+
+  get filtersConfig () {
```
Comment (brennancheung) [inline src/app/plugins/kubernetes/components/apps/AppCatalogPage.js]: Could be much shorter without the getter.

`getFiltersConfig = () => [{ ... }]`

File: src/app/plugins/openstack/index.js [hunk 7]
```diff
@@ -1,57 +1,56 @@
...
 
+import ApiAccessPage from './components/api-access/ApiAccessListPage'
+import AppCatalogPage from 'k8s/components/apps/AppCatalogPage'
+
 import DashboardPage from './components/DashboardPage'
+import AddFlavorPage from './components/flavors/AddFlavorPage'
 
-import AddTenantPage from './components/tenants/AddTenantPage'
-import TenantsListPage from './components/tenants/TenantsListPage'
+import FlavorsListPage from './components/flavors/FlavorsListPage'
```
Comment (brennancheung) [inline src/app/plugins/openstack/index.js]: Let's keep the imports organized by topic.

File: src/server/presets/dev.js [hunk 8]
```diff
@@ -124,37 +127,41 @@ function loadPreset () {
...
-  // Nodes
-  // Nodes must be linked to a resMgrHost id or else the UI will break
-  const node = Node.create({
-    data: { name: 'fakeNode1', api_responding: 1, isMaster: 1, uuid: resMgrHost.id },
-    context,
-    raw: true
-  })
-  Node.create({ data: { name: 'fakeNode2', uuid: resMgrHost2.id }, context })
+  range(4).forEach(clusterIdx => {
+    const cluster = Cluster.create({ data: { name: `fakeCluster${clusterIdx}`, sshKey: 'someKey' }, context, raw: true })
```
Comment (brennancheung) [inline src/server/presets/dev.js]: Please verify the Cypress integration tests for `clusters.spec.js` still work.  We might need to start splitting the presets into smaller chunks and loading multiple of them.

File: src/stories/common/CardTable.stories.js [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+    return this.props.children({
+      filtersConfig,
+      sortingConfig,
+      data: filteredData,
+      classes
+    })
+  }
+}
+
+const DefaultListTable = ({ actions }) => (
```
Comment (brennancheung) [inline src/stories/common/CardTable.stories.js]: Should this be its own component outside of stories?

File: src/stories/common/CardTable.stories.js [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+  actions: {
+    display: 'flex',
+    justifyContent: 'flex-end',
+    alignItems: 'center',
+    padding: theme.spacing.unit
+  }
+})
+
+@withStyles(styles)
+class CardTableContainer extends React.Component {
```
Comment (brennancheung) [inline src/stories/common/CardTable.stories.js]: Should this be its own component outside of stories?

File: src/app/plugins/kubernetes/components/apps/AppCatalogPage.js (hunk 6 above)
Comment (xantorres) [reply src/app/plugins/kubernetes/components/apps/AppCatalogPage.js]: this needs to be called every time as it uses `this.props.context.clusters` which could have changed its reference

File: src/stories/common/CardTable.stories.js (hunk 9 above)
Comment (xantorres) [reply src/stories/common/CardTable.stories.js]: We will see in future implementations if its worth relying on a higher level component like this one that renders the most common implementation of the CardTable and CardItems.
